File: packages/sdk/dart/packages/admin/lib/src/vectorize_client.dart
/// VectorizeClient — Vectorize index access for server-side use.
///
/// Note: Vectorize is Edge-only. In local/Docker, the server returns stub responses.
///
/// ```dart
/// await admin.vector('embeddings').upsert([
///   {'id': 'doc-1', 'values': [0.1, 0.2, 0.3]},
/// ]);
/// final results = await admin.vector('embeddings').search([0.1, 0.2], topK: 5);
/// ```
import 'package:edgebase_core/src/http_client.dart';
import 'generated/admin_api_core.dart';

/// Client for a user-defined Vectorize index.
class VectorizeClient {
  final HttpClient _http;
  final GeneratedAdminApi _core;
  final String _index;

  VectorizeClient(this._http, this._index) : _core = GeneratedAdminApi(_http);

  /// Insert or update vectors.
  /// Returns mutation result with `ok`, optional `count` and `mutationId`.
  Future<Map<String, dynamic>> upsert(List<Map<String, dynamic>> vectors) async {
    final res = await _core.vectorizeOperation(_index, {
      'action': 'upsert',
      'vectors': vectors,
    });
    return res is Map<String, dynamic> ? res : {};
  }

  /// Insert vectors (errors on duplicate ID — server returns 409).
  /// Returns mutation result with `ok`, optional `count` and `mutationId`.
  Future<Map<String, dynamic>> insert(List<Map<String, dynamic>> vectors) async {
    final res = await _core.vectorizeOperation(_index, {
      'action': 'insert',
      'vectors': vectors,
    });
    return res is Map<String, dynamic> ? res : {};
  }

  /// Search for similar vectors.
  Future<List<Map<String, dynamic>>> search(
    List<double> vector, {
    int topK = 10,
    Map<String, dynamic>? filter,
    String? namespace,
    bool? returnValues,
    String? returnMetadata,
  }) async {
    final body = <String, dynamic>{
      'action': 'search',
      'vector': vector,
      'topK': topK,
    };
    if (filter != null) body['filter'] = filter;
    if (namespace != null) body['namespace'] = namespace;
    if (returnValues != null) body['returnValues'] = returnValues;
    if (returnMetadata != null) body['returnMetadata'] = returnMetadata;
    final res = await _core.vectorizeOperation(_index, body);
    if (res is Map && res['matches'] is List) {
      return (res['matches'] as List).cast<Map<String, dynamic>>();
    }
    return [];
  }

  /// Search by an existing vector's ID (Vectorize v2 only).
  Future<List<Map<String, dynamic>>> queryById(
    String vectorId, {
    int topK = 10,
    Map<String, dynamic>? filter,
    String? namespace,
    bool? returnValues,
    String? returnMetadata,
  }) async {
    final body = <String, dynamic>{
      'action': 'queryById',
      'vectorId': vectorId,
      'topK': topK,
    };
    if (filter != null) body['filter'] = filter;
    if (namespace != null) body['namespace'] = namespace;
    if (returnValues != null) body['returnValues'] = returnValues;
    if (returnMetadata != null) body['returnMetadata'] = returnMetadata;
    final res = await _core.vectorizeOperation(_index, body);
    if (res is Map && res['matches'] is List) {
      return (res['matches'] as List).cast<Map<String, dynamic>>();
    }
    return [];
  }

  /// Retrieve vectors by their IDs.
  Future<List<Map<String, dynamic>>> getByIds(List<String> ids) async {
    final res = await _core.vectorizeOperation(_index, {
      'action': 'getByIds',
      'ids': ids,
    });
    if (res is Map && res['vectors'] is List) {
      return (res['vectors'] as List).cast<Map<String, dynamic>>();
    }
    return [];
  }

  /// Delete vectors by IDs.
  /// Returns mutation result with `ok`, optional `count` and `mutationId`.
  Future<Map<String, dynamic>> delete(List<String> ids) async {
    final res = await _core.vectorizeOperation(_index, {
      'action': 'delete',
      'ids': ids,
    });
    return res is Map<String, dynamic> ? res : {};
  }

  /// Get index info (vector count, dimensions, metric).
  Future<Map<String, dynamic>> describe() async {
    final res = await _core.vectorizeOperation(_index, {
      'action': 'describe',
    });
    return res is Map<String, dynamic> ? res : {};
  }
}

File: packages/sdk/csharp/packages/admin/VectorizeClient.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using EdgeBase.Generated;
// EdgeBase C# Admin SDK — Vectorize Client
// Cloudflare Vectorize 인덱스 접근. Service Key 전용.
// All HTTP calls delegate to Generated Admin Core (GeneratedAdminApi).
// No hardcoded API paths — the core is the single source of truth.

namespace EdgeBase.Admin
{

/// <summary>Vectorize 인덱스 클라이언트.</summary>
public sealed class VectorizeClient
{
    private readonly JbHttpClient _http;
    private readonly GeneratedAdminApi _adminCore;
    private readonly string _index;

    internal VectorizeClient(JbHttpClient http, string index)
    {
        _http      = http;
        _adminCore = new GeneratedAdminApi(http);
        _index     = index;
    }

    /// <summary>벡터 Upsert (삽입/갱신).</summary>
    public Task<Dictionary<string, object?>> UpsertAsync(
        VectorInput[] vectors, CancellationToken ct = default)
    {
        var body = new Dictionary<string, object?>
        {
            ["action"]  = "upsert",
            ["vectors"] = vectors
        };
        return _adminCore.VectorizeOperationAsync(Uri.EscapeDataString(_index), body, ct);
    }

    /// <summary>벡터 Insert (중복 ID 시 409 에러).</summary>
    public Task<Dictionary<string, object?>> InsertAsync(
        VectorInput[] vectors, CancellationToken ct = default)
    {
        var body = new Dictionary<string, object?>
        {
            ["action"]  = "insert",
            ["vectors"] = vectors
        };
        return _adminCore.VectorizeOperationAsync(Uri.EscapeDataString(_index), body, ct);
    }

    /// <summary>유사도 검색.</summary>
    /// <param name="vector">쿼리 벡터</param>
    /// <param name="topK">반환할 최대 결과 수</param>
    /// <param name="filter">메타데이터 필터 (선택)</param>
    /// <param name="ns">네임스페이스 (선택)</param>
    /// <param name="returnValues">벡터 값 반환 여부 (선택)</param>
    /// <param name="returnMetadata">메타데이터 반환 수준: all, indexed, none (선택)</param>
    public async Task<List<VectorMatch>> SearchAsync(
        double[] vector, int topK = 10,
        Dictionary<string, object?>? filter = null,
        string? ns = null,
        bool? returnValues = null,
        string? returnMetadata = null,
        CancellationToken ct = default)
    {
        var body = new Dictionary<string, object?>
        {
            ["action"] = "search",
            ["vector"] = vector,
            ["topK"]   = topK
        };
        if (filter != null) body["filter"] = filter;
        if (ns != null) body["namespace"] = ns;
        if (returnValues != null) body["returnValues"] = returnValues;
        if (returnMetadata != null) body["returnMetadata"] = returnMetadata;

        var result = await _adminCore.VectorizeOperationAsync(
            Uri.EscapeDataString(_index), body, ct);
        return JsonHelper.ExtractVectorMatches(result, "matches");
    }

    /// <summary>기존 벡터 ID로 유사도 검색 (Vectorize v2 전용).</summary>
    public async Task<List<VectorMatch>> QueryByIdAsync(
        string vectorId, int topK = 10,
        Dictionary<string, object?>? filter = null,
        string? ns = null,
        bool? returnValues = null,
        string? returnMetadata = null,
        CancellationToken ct = default)
    {
        var body = new Dictionary<string, object?>
        {
            ["action"]   = "queryById",
            ["vectorId"] = vectorId,
            ["topK"]     = topK
        };
        if (filter != null) body["filter"] = filter;
        if (ns != null) body["namespace"] = ns;
        if (returnValues != null) body["returnValues"] = returnValues;
        if (returnMetadata != null) body["returnMetadata"] = returnMetadata;

        var result = await _adminCore.VectorizeOperationAsync(
            Uri.EscapeDataString(_index), body, ct);
        return JsonHelper.ExtractVectorMatches(result, "matches");
    }

    /// <summary>ID로 벡터 조회.</summary>
    public async Task<List<VectorResult>> GetByIdsAsync(
        string[] ids, CancellationToken ct = default)
    {
        var body = new Dictionary<string, object?>
        {
            ["action"] = "getByIds",
            ["ids"]    = ids
        };
        var result = await _adminCore.VectorizeOperationAsync(
            Uri.EscapeDataString(_index), body, ct);
        return JsonHelper.ExtractVectorResults(result, "vectors");
    }

    /// <summary>벡터 삭제.</summary>
    public Task<Dictionary<string, object?>> DeleteAsync(
        string[] ids, CancellationToken ct = default)
    {
        var body = new Dictionary<string, object?>
        {
            ["action"] = "delete",
            ["ids"]    = ids
        };
        return _adminCore.VectorizeOperationAsync(Uri.EscapeDataString(_index), body, ct);
    }

    /// <summary>인덱스 정보 조회 (벡터 수, 차원, 메트릭).</summary>
    public async Task<IndexInfo> DescribeAsync(CancellationToken ct = default)
    {
        var body = new Dictionary<string, object?>
        {
            ["action"] = "describe"
        };
        var result = await _adminCore.VectorizeOperationAsync(
            Uri.EscapeDataString(_index), body, ct);
        return JsonHelper.ExtractIndexInfo(result);
    }
}

/// <summary>Upsert용 벡터 입력.</summary>
public sealed class VectorInput
{
    public string Id { get; set; } = "";
    public double[] Values { get; set; } = Array.Empty<double>();
    public Dictionary<string, object?>? Metadata { get; set; }
    public string? Namespace { get; set; }
}

/// <summary>검색 결과 벡터 매치.</summary>
public sealed class VectorMatch
{
    public string Id { get; set; } = "";
    public double Score { get; set; }
    public double[]? Values { get; set; }
    public Dictionary<string, object?>? Metadata { get; set; }
    public string? Namespace { get; set; }
}

/// <summary>getByIds 결과 벡터.</summary>
public sealed class VectorResult
{
    public string Id { get; set; } = "";
    public double[]? Values { get; set; }
    public Dictionary<string, object?>? Metadata { get; set; }
    public string? Namespace { get; set; }
}

/// <summary>describe 결과 인덱스 정보.</summary>
public sealed class IndexInfo
{
    public long VectorCount { get; set; }
    public int Dimensions { get; set; }
    public string Metric { get; set; } = "";
    public string? Id { get; set; }
    public string? Name { get; set; }
}
}
